File: XPC Services/Inline Content Loader/Classes/Modules/Twitter/ICMTweet.m
/* ********************************************************************* 
                  _____         _               _
                 |_   _|____  _| |_ _   _  __ _| |
                   | |/ _ \ \/ / __| | | |/ _` | |
                   | |  __/>  <| |_| |_| | (_| | |
                   |_|\___/_/\_\\__|\__,_|\__,_|_|

 Copyright (c) 2010 - 2017 Codeux Software, LLC & respective contributors.
        Please see Acknowledgements.pdf for additional information.

 Redistribution and use in source and binary forms, with or without
 modification, are permitted provided that the following conditions
 are met:

    * Redistributions of source code must retain the above copyright
      notice, this list of conditions and the following disclaimer.
    * Redistributions in binary form must reproduce the above copyright
      notice, this list of conditions and the following disclaimer in the
      documentation and/or other materials provided with the distribution.
    * Neither the name of Textual and/or "Codeux Software, LLC", nor the 
      names of its contributors may be used to endorse or promote products 
      derived from this software without specific prior written permission.

 THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
 ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 ARE DISCLAIMED. IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
 FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 SUCH DAMAGE.

 *********************************************************************** */

#import "ICLInlineContentModulePrivate.h"
#import "ICLHelpers.h"
#import "ICMTweet.h"

NS_ASSUME_NONNULL_BEGIN

@implementation ICMTweet

- (void)_loadTweetContents
{
	NSString *tweetAddress = self.payload.address;

	NSURLComponents *requestComponents = [NSURLComponents componentsWithString:@"https://publish.twitter.com/oembed"];

	requestComponents.queryItems =
	@[
	  [NSURLQueryItem queryItemWithName:@"dnt" value:@"true"], /* DO NOT TRACK */
	  [NSURLQueryItem queryItemWithName:@"maxwidth" value:@"500"],
	  [NSURLQueryItem queryItemWithName:@"omit_script" value:@"true"],
	  [NSURLQueryItem queryItemWithName:@"url" value:tweetAddress]
	];

	NSURL *requestURL = requestComponents.URL;

	[ICLHelpers requestJSONObject:@"html"
						   ofType:[NSString class]
					  inHierarchy:nil
						  fromURL:requestURL
				  completionBlock:^(id object) {
				if (object == nil) {
					[self notifyUnableToPresentHTML];

					return;
				}

				[self performActionForHTML:object];
			}];
}

#pragma mark -
#pragma mark Action Block

+ (nullable SEL)actionForURL:(NSURL *)url
{
	NSParameterAssert(url != nil);

	if ([self _URLIsTweet:url] == NO) {
		return NULL;
	}

	return @selector(_loadTweetContents);
}

+ (BOOL)_URLIsTweet:(NSURL *)url
{
	NSString *urlPath = url.path.percentEncodedURLPath;

	if (urlPath.length == 0) {
		return NO;
	}

	urlPath = [urlPath substringFromIndex:1]; // "/"

	NSArray<NSString *> *components = [urlPath componentsSeparatedByString:@"/"];

	if (components.count < 3) {
		return NO;
	}

	if ([components[1] isEqualToString:@"status"] == NO) {
		return NO;
	}

	if (components[2].isNumericOnly == NO) {
		return NO;
	}

	return YES;
}

+ (nullable NSArray<NSString *> *)domains
{
	static NSArray<NSString *> *domains = nil;

	static dispatch_once_t onceToken;

	dispatch_once(&onceToken, ^{
		domains =
		@[
		  @"twitter.com",
		  @"www.twitter.com",
		  @"mobile.twitter.com"
		];
	});

	return domains;
}

#pragma mark -
#pragma mark Utilities

- (nullable NSArray<NSURL *> *)scriptResources
{
	return
	[[super scriptResources] arrayByAddingObjectsFromArray:
	@[
	  [NSURL URLWithString:@"https://platform.twitter.com/widgets.js"],
	  [RZMainBundle() URLForResource:@"ICMTweet" withExtension:@"js" subdirectory:@"Components"]
	]];
}

- (nullable NSString *)entrypoint
{
	return @"_ICMTweet";
}

- (void)finalizePreflight
{
	self.payload.classAttribute = @"inlineTweet";
}

@end

NS_ASSUME_NONNULL_END

File: XPC Services/Inline Content Loader/Extensions/Core Media/Resources/Modules/Twitter/ICMTweet.js
/* *********************************************************************
 *                  _____         _               _
 *                 |_   _|____  _| |_ _   _  __ _| |
 *                   | |/ _ \ \/ / __| | | |/ _` | |
 *                   | |  __/>  <| |_| |_| | (_| | |
 *                   |_|\___/_/\_\\__|\__,_|\__,_|_|
 *
 * Copyright (c) 2017, 2018 Codeux Software, LLC & respective contributors.
 *       Please see Acknowledgements.pdf for additional information.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 *
 *  * Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 *  * Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 *  * Neither the name of Textual, "Codeux Software, LLC", nor the
 *    names of its contributors may be used to endorse or promote products
 *    derived from this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 *
 *********************************************************************** */

window.twttr = (function() { /* cancer */
	var t = window.twttr;

	if (t) {
		return t;
	}

	t = {};

	t._e = [];

	t.ready = function(f) {
		t._e.push(f);
	};

	return t;
}());

var _ICMTweetPrototypeParent = _ICMInlineHTMLPrototype;

var _ICMTweetPrototype = function() {
	_ICMTweetPrototypeParent.call(this);
}

_ICMTweetPrototype.prototype = Object.create(_ICMInlineHTMLPrototype.prototype);
_ICMTweetPrototype.prototype.constructor = _ICMTweetPrototype;
_ICMTweetPrototype.prototype.superClass = _ICMTweetPrototypeParent.prototype;

_ICMTweetPrototype.prototype.didLoadMedia = function(mediaId, mediaElement)
{
	/* We do not have enough context in the module itself to set 
	the appearance of it, but we do once we reach the entrypoint. */
	var themeAppearance = document.body.dataset.appearance;

	if (themeAppearance === "dark") {
		var tweet = mediaElement.querySelector("blockquote.twitter-tweet");

		tweet.dataset.theme = "dark";
	}

	twttr.widgets.load(mediaElement);
};

_ICMTweetPrototype.prototype.tweetRendered = function(event)
{
	TextualScroller.restoreScrolledToBottom();
};

var _ICMTweet = null;

twttr.ready(
	function(twttr) {
		_ICMTweet = new _ICMTweetPrototype();

		twttr.events.bind("rendered", _ICMTweet.tweetRendered);
	}
);
